core/src/main/java/io/grpc/util/RoundRobinLoadBalancerFactory2.java

```diff
@@ -0,0 +1,259 @@
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.util;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.grpc.ConnectivityState.IDLE;
+import static io.grpc.ConnectivityState.TRANSIENT_FAILURE;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.BiMap;
+import com.google.common.collect.HashBiMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+
+import io.grpc.Attributes;
+import io.grpc.ConnectivityState;
+import io.grpc.ConnectivityStateInfo;
+import io.grpc.EquivalentAddressGroup;
+import io.grpc.ExperimentalApi;
+import io.grpc.LoadBalancer;
+import io.grpc.LoadBalancer2;
+import io.grpc.LoadBalancer2.PickResult;
+import io.grpc.LoadBalancer2.Subchannel;
+import io.grpc.LoadBalancer2.SubchannelPicker;
+import io.grpc.Metadata;
+import io.grpc.NameResolver;
+import io.grpc.ResolvedServerInfo;
+import io.grpc.ResolvedServerInfoGroup;
+import io.grpc.Status;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * A {@link LoadBalancer} that provides round-robin load balancing mechanism over the
+ * addresses from the {@link NameResolver}.  The sub-lists received from the name resolver
+ * are considered to be an {@link EquivalentAddressGroup} and each of these sub-lists is
+ * what is then balanced across.
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/1771")
+public class RoundRobinLoadBalancerFactory2 extends LoadBalancer2.Factory {
+  private static final RoundRobinLoadBalancerFactory2 instance =
+      new RoundRobinLoadBalancerFactory2();
+
+  private RoundRobinLoadBalancerFactory2() {
+  }
+
+  public static RoundRobinLoadBalancerFactory2 getInstance() {
+    return instance;
+  }
+
+  @Override
+  public LoadBalancer2 newLoadBalancer(LoadBalancer2.Helper helper) {
+    return new RoundRobinLoadBalancer(helper);
+  }
+
+  @VisibleForTesting
+  static class RoundRobinLoadBalancer extends LoadBalancer2 {
+    private final Helper helper;
+    private final BiMap<EquivalentAddressGroup, Subchannel> subchannels = HashBiMap.create();
```

Comment:
There is one thing to note that this will not keep the original order and any duplicate entries from NameResolver.
Not sure if this matters, but this is a behavioral difference from the original implementation.
Might consider having a `List<EquivalentAddressGroup>` to keep the original list and use it for constructing the list in the picker.